### DISCUSSION TITLE
Slowly leaking connections - How to get a grip?

### DISCUSSION THREAD

**Discussion by ErrorProne** [quarkusio/quarkus#27566] [Q&A]
Hi everyone,

over the past months we've built an application which controls an assembly line. We are now slowly starting into production and  recognized that we leak DB connections and I can't get a grip on it.

![Screenshot 2022-08-29 at 11 21 59](https://user-images.githubusercontent.com/541410/187169896-6cba7d52-45fc-4508-96db-75132fe825b9.png)
The blue line is the "agroal_active_count" and green "agroal_available_count" from the metrics extension. The cut on the 27. is me restarting the application.

What I've seen so far:
- It looks like the DB pool is very slowly not reusing connections, since the active count is increasing despite the fact that there are available connections.
- I would have expected the active count to decrease over the night. The available count matches the working hours in the factory pretty good and after that should the DB pool not release some of the connections?
- Looking at postgres activity output I can see that we get more and more connections but they are all idle and not associated with any transaction or active query.
- We have not a single leaked connection warning in the log and all method calls finish in <1s
- We only use @Transactional and in some places `QuarkusTransaction.run()`

Our datasource configuration is pretty straight forward and default

```yaml
  datasource:
    db-kind: postgresql
    jdbc:
      url: "jdbc:postgresql://localhost:5432/wcs?assumeMinServerVersion=9.0"
      max-size: 100
```

Does anyone have any pointers or ideas what settings to tune to get more information about what is going on?

Environment:
- quarkus 2.10.2.Final
- openJDK 17.0.4 (with ZGC)
- PostgreSQL 14
- UberJar-Deployment

**Comment by Sanne** [comment]
You have configured `max-size: 100`, in this case Agroal will slowly build up the number of active connections up to having 100.
So I'm not sure if this is a "leak" - but it seems to grow suspiciously indeed.

Is it possible to verify how your system would behave if you were to set a lower number? Is it ever going to stabilize at the maximum, or does it get in trouble?

> It looks like the DB pool is very slowly not reusing connections, since the active count is increasing despite the fact that there are available connections.

How do you know there are "available connections"? the green line on your plots look very low, close to zero in many instants, so I'd expect it indeed to need to enlarge the pool further.

Are you sure there are no long running processes that are actively using them?

**Reply by ErrorProne** [reply]
**This thread lead to some kind of rubberducking, one can skip to the third post for the results**

> Is it possible to verify how your system would behave if you were to set a lower number? Is it ever going to stabilize at the maximum, or does it get in trouble?

We are currently trying to reproduce it somehow on our dev environment (we something file style playbooks to simulate a production day), but without real success so far but it will be running for a few more hours so we will see.  We've tipped the limit in production once, which caused the machines to stop so I'm not eager to try it out again. The postgres internal max connection limit was set too low before, so it clashed with the 100 client connections and the 3 reserved for admins. However, even when we where only 3 connections short the application blocked up severely.

> How do you know there are "available connections"? the green line on your plots look very low, close to zero in many instants, so I'd expect it indeed to need to enlarge the pool further.

I've queried the pg_stat_activity first for inactive connection. 
Something like `SELECT * from pg_stat_activity WHERE state in ('idle', 'idle in transaction', 'idle in transaction (aborted)', 'disabled')  AND current_timestamp - state_change > interval '5 minutes' ;` 
Which gave me a > 50 connections back than. Also right now (restarted) today we have currently 22 total connections and 16 of them "inactive" and some of then have their last state_change yesterday.

> Are you sure there are no long running processes that are actively using them?

Pretty much, yes. We have to react to events coming in from machines and give a response to the control unit for further actions and don't have much time (mostly <1s at some points <200ms) for this since they conveyor belts don't wait for us. This is also way we have ZGC in place, otherwise the GC pauses would sometimes be too long for this usecase.
That being said, we had problems with some statistic queries in the (very small) ui, but that has been fixed as far as I can tell. Also, with long running processes I would have suspected that they will show up in postgres activity table as idle in transaction or in transaction.

> You have configured max-size: 100, in this case Agroal will slowly build up the number of active connections up to having 100.
So I'm not sure if this is a "leak" - but it seems to grow suspiciously indeed.

Coming back to this, I would have guessed that the pool only acquires new connections only if required, but reading what you've said I guess the logic and performance tweaks behind that are much more complex.
I was so focused on the active count that I didn't saw the low number of available connections which does not match what I see in postgres. Just now I see in the agroal stats 16 active connections and 3 available. Looking at postgres I've 16 which I would call inactive (state idle and no state change for more than 5 minutes) and 19 connections in total.

Maybe I've just misinterpreted the term active? Active for me is a connection actively used in a transaction or query.

**Reply by ErrorProne** [reply]
I've added the rate calculated by grafana/prometheus for the agroal_acquire_count_total value which gives me a chart like this for the last 12hours

![Screenshot 2022-08-30 at 15 01 18](https://user-images.githubusercontent.com/541410/187443332-b8942d78-e6c0-4ea6-a679-bf9317d6b8b5.png)

As I would interpret it the pool creates a lot of new connections which is totally fine, leaves some available connections. 

Looking at this, and knowing how the application work, it looks to me like all the event driven part from the machines is fine, which is 99% of the work we do.
The statistics in the frontend are fetched via SSE, could it be that we block connections there, which are associated to an SSE stream but not in an active transaction from the postgres point of view? That would also mean, given that some of the connections are from last night which would matches with my time checking the application after the restart, that connections used in that particular case are not released.
Maybe I've missed something on connection handling in SSE.

**Reply by ErrorProne** [reply]
Update: Testing with the SSE idea we've narrowed it down to those calls and found a service, which is called in each tick of the multi to query the charts, and did not have an `@Transactional`. Reloading the site somehow leaves the connection as active in the pool (wild guess: it is associated to some thread?). This has been fixed by adding `@Transactional`. 

Interestingly the behaviour of available connections changed completely

![Screenshot 2022-08-30 at 16 17 36](https://user-images.githubusercontent.com/541410/187461459-9e350994-e329-4e2e-8aaf-3bf9fe389210.png)

The red line in the middle marks a redeploy. On the left we had a deployment running with only an additional `@Transactional` on the service (fixed version) and on the right the same without (buggy version).
The REST endpoint is more or less a simple

```java
        ...
	@RestStreamElementType("text/vnd.turbo-stream.html")
	public Multi<String> all(@QueryParam("chartDate") Instant chartDate) {
            ...
	    return createMultiForRefreshInterval()
		    .onItem()
		    .transform((Long tick) -> getChartsForDate(finalizedChartDate));
	}
```

What intrigues me, and maybe you have some pointers for me, is that with the `@Transactional` at the service level, for the chart generation, on the left the pool leaves a constant supply of 25 available connections while with the buggy version this is not the case.

**Reply by ErrorProne** [reply]
I'm still on my journey to pin this further down. It is somehow related to the SSE endpoint if the request is cancelled. But I've found a second endpoint which did something similar and did not have the problem that connections remained as active in the pool.
So it looks like there is more to it

**Reply by Sanne** [reply]
I suspect you're hitting some limitations in mixing reactive programming with the blocking JDBC datasource; in particular it's a bit tricky to define were a scope is terminated when having a reactive stream, but I'd love to be able to reproduce it to see if we can do anything to make this easier.

Do you think you could narrow it down to a reproducer? I suspect this relates to how exactly the JDBC datasource is accessed from the reactive stream - indeed forgetting `@Transactional` or otherwise marking it as a blocking operation would result in erroneus behaviour; I need to ensure if we can make this less errorprone.

**Reply by Sanne** [reply]
I don't think it's the same issue, but there's some similarities with https://github.com/quarkusio/quarkus/issues/27166

**Reply by ErrorProne** [reply]
> Do you think you could narrow it down to a reproducer?

I'll try to extract the essential part to reproduce it. But I'm not sure if I've time for it today.

**Reply by Sanne** [reply]
Thanks! and no rush
